clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -221,8 +222,45 @@ public int partitionLeaderEpoch() {
         return buffer.getInt(PARTITION_LEADER_EPOCH_OFFSET);
     }
 
+    public boolean validateBatchSize() {
```

Comment:
Hmm... Seems rather wasteful to decompress another time just to validate the record count. What I had thought is that we could hook the logic into `DefaultRecordBatch.RecordIterator` so that we can do this validation as part of the single pass we do over the records when validating. As a matter of fact, we seem to already do part of the validation there. If the number of records that we consumed is greater than the count, then we raise `NoSuchElementException`. That should probably be changed to `InvalidRecordException` and then we just need to check for the underflow case.